### PR TITLE
Make import/export devices tests more resilient

### PR DESCRIPTION
**Pull request by drwill-ms**
Includes adding a convenience property to the JobProperties object. Thoughts?

### CODE REVIEW

**Review by timtay-microsoft — APPROVED** [review]
I like the new property on JobProperties here

File: iothub/service/src/Jobs/JobProperties.cs [hunk 1]
```diff
@@ -143,6 +151,12 @@ public JobProperties()
         [JsonProperty(PropertyName = "configurationsBlobName", NullValueHandling = NullValueHandling.Ignore)]
         public string ConfigurationsBlobName { get; set; }
 
+        /// <summary>
+        /// Convenience property to determine if the job is in a terminal state, based on <see cref="JobStatus"/>.
+        /// </summary>
+        [JsonIgnore]
+        public bool IsFinished => s_finishedStates.Contains(Status);
```
Comment (drwill-ms) [inline iothub/service/src/Jobs/JobProperties.cs:158]: IsComplete? IsTerminal?

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs [hunk 2]
```diff
@@ -163,55 +155,57 @@ public async Task RegistryManager_ExportDevices(StorageAuthenticationType storag
...
-                    exportJobResponse = await registryManager.ExportDevicesAsync(exportJobResponse).ConfigureAwait(false);
+                    jobProperties = await registryManager.ExportDevicesAsync(jobProperties).ConfigureAwait(false);
+                    if (!string.IsNullOrWhiteSpace(jobProperties.FailureReason))
+                    {
+                        Logger.Trace($"Job failed due to {jobProperties.FailureReason}");
+                    }
                     break;
                 }
                 // Concurrent jobs can be rejected, so implement a retry mechanism to handle conflicts with other tests
-                catch (JobQuotaExceededException) when (++tryCount < MaxIterationWait)
```
Comment (drwill-ms) [inline e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:192]: Rely on the test timeout, not an arbitrary number of retries (which when multiplied by the sleep gives a not-obvious test timeout).

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs [hunk 3]
```diff
@@ -163,55 +155,57 @@ public async Task RegistryManager_ExportDevices(StorageAuthenticationType storag
...
-                    Logger.Trace($"JobQuotaExceededException... waiting.");
+                    Logger.Trace($"JobQuotaExceededException... waiting after {sw.Elapsed}.");
                     await Task.Delay(s_waitDuration).ConfigureAwait(false);
                     continue;
                 }
             }
 
-            for (int i = 0; i < MaxIterationWait; ++i)
+            sw.Stop();
+            Logger.Trace($"Job started after {sw.Elapsed}.");
```
Comment (drwill-ms) [inline e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:197]: Print out some time taken to start the job or wait for it to finish so we get an idea of how long we're waiting if a test times out.

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs [hunk 4]
```diff
@@ -313,13 +307,37 @@ private static async Task<string> DownloadFileAsync(StorageContainer storageCont
             try
             {
                 await registryManager.RemoveDeviceAsync(deviceId).ConfigureAwait(false);
+            }
```
Comment (drwill-ms) [inline e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:310]: Give each device/config a chance to get deleted.

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs [hunk 5]
```diff
@@ -313,13 +307,37 @@ private static async Task<string> DownloadFileAsync(StorageContainer storageCont
...
+                Logger.Trace($"Failed to remove device during cleanup due to {ex.Message}");
+            }
+
+            try
+            {
                 await registryManager.RemoveDeviceAsync(edgeId2).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Logger.Trace($"Failed to remove device during cleanup due to {ex.Message}");
```
Comment (tmahmood-microsoft) [inline e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:322]: how would we know which device failed?

File: iothub/service/src/Jobs/JobProperties.cs (hunk 1 above)
Comment (abhipsaMisra) [reply iothub/service/src/Jobs/JobProperties.cs:158]: Given that "Complete" is its own status, I like "IsFinished".

**Review by abhipsaMisra — APPROVED** [review]

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs [hunk 6]
```diff
@@ -313,13 +307,37 @@ private static async Task<string> DownloadFileAsync(StorageContainer storageCont
             try
             {
                 await registryManager.RemoveDeviceAsync(deviceId).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+            {
+                Logger.Trace($"Failed to remove device during cleanup due to {ex.Message}");
```
Comment (abhipsaMisra) [inline e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:313]:
```suggestion
                Logger.Trace($"Failed to remove device {deviceId} during cleanup due to {ex.Message}");
```

File: e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs (hunk 5 above)
Comment (abhipsaMisra) [reply e2e/test/iothub/service/RegistryManagerExportDevicesTests.cs:322]:
```suggestion
                Logger.Trace($"Failed to remove device {edgeId2} during cleanup due to {ex.Message}");
```

**Review by tmahmood-microsoft — APPROVED** [review]

**Pull request merged**